### PR TITLE
Prometheus collector

### PR DESCRIPTION
**Pull request by vkhromov**
Example of configuration:
```json
"endpoints": [
    {
        "prefix":         "etcd/",
        "url":            "https://etcd1.nowhere.com:2379/metrics",
        "serverCaFile":   "/etc/pki/server-ca.crt",
        "clientCertFile": "/etc/pki/client-etcd1-nowhere.com.crt",
        "clientKeyFile":  "/etc/pki/client-etcd1-nowhere.com.key",
        "metrics_whitelist": [
            "123",
            "456"
        ],
        "metrics_blacklist": {
            "78"
        },
        "generated_dimensions": {
            "hostname": "etcd1.nowhere.com",
        }
    }
]
```

### CODE REVIEW

**Comment by EmanekaT** [comment]
Looks good overall. I'm thinking if we can make it as a library instead. It seems the scape and extractMetric functions are highly reusable.

File: src/fullerite/collector/prometheus.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+// Collect iterates on all Prometheus endpoints and collect the corresponding metrics
+// For each endpoint a gorutine is started to spin up the collection process.
+func (p *Prometheus) Collect() {
+	for _, endpoint := range p.endpoints {
+		go p.collectFromEndpoint(endpoint)
+	}
+}
+
+const (
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: Could you add some comments to explain?

File: src/fullerite/collector/prometheus.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+	p.name = "Prometheus"
+	return p
+}
+
+// Configure takes a dictionary of values with which the handler can configure itself.
+func (p *Prometheus) Configure(configMap map[string]interface{}) {
+	if v, exists := configMap["endpoints"]; !exists {
+		p.log.Fatal("No endpoints specified in Prometheus config")
+	} else {
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: Nit: Perhaps remove else ? Nested if else is not very readable.

File: src/fullerite/collector/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+		if endpoints, ok := v.([]interface{}); !ok {
+			p.log.Fatal("Invalid format of config entry `endpoints'")
+		} else {
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: Same

File: src/fullerite/collector/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+			for _, e := range endpoints {
+				if endpoint, ok := e.(map[string]interface{}); !ok {
+					p.log.Fatal("Invalid format of an item in config entry `endpoints`")
+				} else {
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: same

File: src/fullerite/collector/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+							p.log.Fatalf("Invalid format of config entry `generated_dimensions'")
+						} else {
+							generatedDimensions = v2
+						}
+					}
+
+					metricsWhitelist := p.getSet(endpoint, "metrics_whitelist")
+					metricsBlacklist := p.getSet(endpoint, "metrics_blacklist")
+
+					var timeout = 5
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: Move 5 to top of the file. var defaultTimeoutSecs int = 5

**Comment by stug** [comment]
I think this also looks good to me but I am new to Go and a not very familiar with Fullerite!

File: src/fullerite/collector/prometheus.go (hunk 2 above)
Comment (vkhromov) [reply src/fullerite/collector/prometheus.go]: please note that using `else` is intentional here:
* `v` is declared in the `if` statement and then is  used in the `else` branch
* I still not sure that using `Fatal` is okey in this scenario

File: src/fullerite/collector/prometheus.go (hunk 1 above)
Comment (vkhromov) [reply src/fullerite/collector/prometheus.go]: :+1:

**Comment by vkhromov** [comment]
@EmanekaT, I've addressed your comment, could you please take another look at the PR?

**Comment by stug** [comment]
Also some Prometheus metrics have additional dimensions on them (e.g. `kubelet_docker_operations{operation_type="pull_image"} 17`) -- does this handle that?  Seems like not?

File: src/fullerite/collector/prometheus.go [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, "", errors.Wrap(err, "Error while reading the response body")
+	}
+
+	return body, resp.Header.Get("Content-Type"), nil
+}
+
+func extractPrometheusMetrics(
```
Comment (EmanekaT) [inline src/fullerite/collector/prometheus.go]: Perhaps putting this in util folder is better? Not a big deal.

**Comment by vkhromov** [comment]
> Also some Prometheus metrics have additional dimensions on them (e.g. `kubelet_docker_operations{operation_type="pull_image"} 17`) -- does this handle that? Seems like not?

@stug please see the update.

**Pull request merged**